### DISCUSSION TITLE
Measure and display heart rate variability (HRV)

### DISCUSSION THREAD

**Discussion by khimaros** [InfiniTimeOrg/InfiniTime#1555] [Ideas]
### Verification

- [X] I searched for similar feature request and found none was relevant.

### Pitch us your idea!

heart rate variability (HRV) measures the dynamic range of your heart

### Description

according to https://my.clevelandclinic.org/health/symptoms/21773-heart-rate-variability-hrv

"If your heart rate is highly variable, this is usually evidence that your body can adapt to many kinds of changes. People with high heart rate variability are usually less stressed and happier."

"In general, low heart rate variability is considered a sign of current or future health problems because it shows your body is less resilient and struggles to handle changing situations."

**Comment by minacode** [comment]
Sounds interesting but the article you provided says
> Outside of a medical setting, there are several devices commonly used by athletes, especially runners. These often consist of a device attached to a band that wraps around your chest. Some look like pulse oximeters (devices that attach to a finger and measure your pulse and blood oxygen level) but are more sensitive and accurate.

and

> The majority of wrist-worn fitness devices and trackers track your heart rate through your skin. Unfortunately, this means they usually aren't sensitive enough to detect heart rate variability accurately.

Feel free to try. But given that I am not sure if we would be able to extract reliable data.

**Comment by khimaros** [comment]
it looks like @Pittconnect made some degree of progress doing HRV analysis with wasp-os and HeartPy: https://github.com/daniel-thompson/wasp-os/issues/273#issuecomment-1002804919 -- here's a video of their attempt: https://www.youtube.com/watch?v=FsX2-1JJF6o

here's a notebook exploring heart rate data from a smart watch with HeartPy: https://github.com/paulvangentcom/heartrate_analysis_python/blob/master/examples/3_smartwatch_data/Analysing_Smartwatch_Data.ipynb

in the example above `rmssd` is the measure of heart rate variability.

there is some additional detail about HRV and its health implications here: https://support.ouraring.com/hc/en-us/articles/360025441974-An-Introduction-to-Heart-Rate-Variability

i think my first step will be getting a few minutes of raw PPG data from the watch and do a bit of offline analysis to see if the signal is clean enough to be useful.

**Comment by minacode** [comment]
This sounds way more promising!

> i think my first step will be getting a few minutes of raw PPG data from the watch and do a bit of offline analysis to see if the signal is clean enough to be useful.

Good luck! Would be nice if you succeed.

**Comment by khimaros** [comment]
now that https://github.com/InfiniTimeOrg/InfiniTime/pull/1486 has been merged, maybe we have reliable enough heart rate data to do something like this.

**Comment by khimaros** [comment]
with the new PPG algorithm, peaks are detected and the amplitude of the previous peak is being saved here: https://github.com/InfiniTimeOrg/InfiniTime/blob/main/src/components/heartrate/Ppg.cpp#L249

it seems like it should be relatively straightforward to also record the time of the previous peak and use that to derive the time between the latest peak and the previous one.

that data could then be fed to the heart rate task and exposed as a GATT characteristic. this way, the HRV (and perhaps other trends) could be calculated off device from the raw peak data.

**Reply by mark9064** [reply]
Unfortunately the peak detection you refer to here is in the frequency domain, not time domain. What's happening here is the algorithm takes a frequency spectrum (like one you might see for music), and then finds an appropriate peak that corresponds to the beat frequency (like pitch for music) of the heart. To do beat-to-beat analysis, a time domain algorithm would be more suitable (or painful frequency domain phase analysis that wouldn't work with the current algorithm implementation)
